hopr/config/src/lib.rs

```diff
@@ -0,0 +1,316 @@
+                winning_probability_oracle: network.addresses.winning_probability_oracle.to_owned(),
+                token: network.addresses.token.to_owned(),
+                tx_polling_interval: network.tx_polling_interval,
+                max_block_range: network.max_block_range,
+                max_requests_per_sec: max_rpc_requests_per_sec.or(chain.max_rpc_requests_per_sec),
+            }),
+            Ok(false) => Err(format!(
+                "network {id} is not supported, supported networks {:?}",
+                protocol_config.supported_networks(version).join(", ")
+            )),
+            Err(e) => Err(e.to_string()),
+        }
+    }
+}
+
+impl From<&ChainNetworkConfig> for ContractAddresses {
+    fn from(network: &ChainNetworkConfig) -> Self {
+        Self {
+            token: network.token,
+            channels: network.channels,
+            announcements: network.announcements,
+            network_registry: network.network_registry,
+            network_registry_proxy: network.network_registry_proxy,
+            safe_registry: network.node_safe_registry,
+            price_oracle: network.ticket_price_oracle,
+            win_prob_oracle: network.winning_probability_oracle,
+            stake_factory: network.node_stake_v2_factory,
+            module_implementation: network.module_implementation,
+        }
+    }
+}
+
+/// The entire on-chain protocol configuration containing the information about
+/// usable networks and chains.
+#[derive(Debug, Clone, Serialize, Deserialize, Validate)]
+#[serde(deny_unknown_fields)]
+pub struct ProtocolsConfig {
+    pub networks: BTreeMap<String, Network>,
+    pub chains: BTreeMap<String, ChainOptions>,
+}
+
+impl Default for ProtocolsConfig {
+    fn default() -> Self {
+        Self::from_str(include_str!("../data/protocol-config.json"))
+            .expect("bundled protocol config should be always valid")
+    }
+}
+
+impl FromStr for ProtocolsConfig {
+    type Err = String;
+
+    /// Reads the protocol config JSON file and returns it
+    fn from_str(data: &str) -> Result<Self, Self::Err> {
+        serde_json::from_str::<ProtocolsConfig>(data).map_err(|e| e.to_string())
+    }
+}
+
+impl std::cmp::PartialEq for ProtocolsConfig {
+    fn eq(&self, other: &Self) -> bool {
+        Vec::from_iter(self.networks.clone()) == Vec::from_iter(other.networks.clone())
+            && Vec::from_iter(self.chains.clone()) == Vec::from_iter(self.chains.clone())
+    }
+}
+
+impl ProtocolsConfig {
+    /// Returns a list of environments which the node is able to work with
+    /// TODO: crate::constants::APP_VERSION_COERCED
+    pub fn supported_networks(&self, version: &str) -> Vec<String> {
+        let mut allowed = vec![];
+
+        for (name, env) in self.networks.iter() {
+            let range = env.version_range.to_owned();
+
+            if let Ok(true) = satisfies(version, range.as_str()) {
+                allowed.push(name.clone())
+            }
+        }
```

Comment:
It'd be preferable to change this in a way, where the user does not specify the network, but rather a protocol config directly. These arbitrary checks would disappear completely.

Reply:
Doing later.